osu.Game/Scoring/ScoreInfoExtensions.cs

```diff
@@ -1,36 +1,161 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
+using osu.Framework.Bindables;
 using osu.Game.Beatmaps;
+using osu.Game.Configuration;
+using osu.Game.Online.API.Requests.Responses;
+using osu.Game.Rulesets;
 using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Scoring
 {
     public static class ScoreInfoExtensions
     {
+        /// <summary>
+        /// Computes the total score based on the given scoring mode.
+        /// </summary>
+        public static long GetDisplayScore(this IScoreInfo score, ScoringMode mode)
+        {
+            switch (score)
+            {
+                case ScoreInfo scoreInfo:
+                    return Legacy.ScoreInfoExtensions.GetDisplayScore(scoreInfo, mode);
+
+                case SoloScoreInfo soloScoreInfo:
+                    return Legacy.ScoreInfoExtensions.GetDisplayScore(soloScoreInfo, mode);
+
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
         /// <summary>
         /// A user-presentable display title representing this score.
         /// </summary>
         public static string GetDisplayTitle(this IScoreInfo scoreInfo) => $"{scoreInfo.User.Username} playing {scoreInfo.Beatmap?.GetDisplayTitle() ?? "unknown"}";
 
         /// <summary>
-        /// Orders an array of <see cref="ScoreInfo"/>s by total score.
+        /// Orders an array of <typeparamref name="TScoreInfo"/>s by total score.
         /// </summary>
-        /// <param name="scores">The array of <see cref="ScoreInfo"/>s to reorder.</param>
+        /// <param name="scores">The array of <typeparamref name="TScoreInfo"/>s to reorder.</param>
         /// <returns>The given <paramref name="scores"/> ordered by decreasing total score.</returns>
-        public static IEnumerable<ScoreInfo> OrderByTotalScore(this IEnumerable<ScoreInfo> scores)
+        public static IEnumerable<TScoreInfo> OrderByTotalScore<TScoreInfo>(this IEnumerable<TScoreInfo> scores)
+            where TScoreInfo : IScoreInfo
             => scores.OrderByDescending(s => s.TotalScore)
                      .ThenBy(s => s.OnlineID)
                      // Local scores may not have an online ID. Fall back to date in these cases.
                      .ThenBy(s => s.Date);
 
+        /// <summary>
+        /// Returns the list of hit result statistics applicable for this score.
+        /// </summary>
+        public static IEnumerable<HitResultDisplayStatistic> GetStatisticsForDisplay(this IScoreInfo score, Ruleset ruleset)
+        {
+            if (ruleset.RulesetInfo.OnlineID != score.Ruleset.OnlineID)
+                throw new InvalidOperationException($@"The ID of the given ruleset instance ({ruleset.RulesetInfo.OnlineID}) does not match the score's ruleset ({score.Ruleset.OnlineID})");
```

Comment:
Have we ever done this before - passing a separate `Ruleset` in? The closest is `SoloScoreInfo` which passes in a `RulesetStore`.

I assume the case here is if someone passes in an `APIRuleset` which doesn't implement `CreateInstance()`. In which case if they use the other overload of this method, things will go ka-boom.

But how about we implement `CreateInstance()`? Just via `AssemblyRulesetStore`. This is exemplified elsewhere:
https://github.com/ppy/osu/blob/8e79ab6c04eb84974b1ee2fe98e6cf8e1a9fe231/osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs#L65-L75

Reply:
> Or - change this to accept a `RulesetStore`, like `SoloScoreInfo`.

In my previous attempts at similar refactors I believe that is what I did.